File: backend/src/orbtronics_l1_software_engineer_backend/schemas/tasks.py
from datetime import date, datetime
from enum import Enum

from pydantic import BaseModel


class TaskPriorityEnum(str, Enum):
    none = "none"
    low = "low"
    medium = "medium"
    high = "high"


class TaskStatusEnum(str, Enum):
    todo = "todo"
    in_progress = "in_progress"
    done = "done"


class TaskDueEnum(Enum):
    none = "none"
    today = "today"
    week = "week"
    overdue = "overdue"


class TaskData(BaseModel):
    task_id: str
    user_id: str
    title: str
    description: str | None = None
    status: str
    priority: str | None = None
    due_date: str | None = None
    created_at: str
    updated_at: str


class TaskDataInput(BaseModel):
    title: str
    description: str | None = None
    status: str = TaskStatusEnum.todo
    priority: str | None = None
    due_date: date | str | None = None


class TaskDataPartial(BaseModel):
    title: str | None
    description: str | None
    status: str | None
    priority: str | None
    due_date: str | None

File: frontend/src/lib/tasks.ts
export type Status = "todo" | "in_progress" | "done";
export type Priority = "high" | "medium" | "low" | "none";

export type TaskInput = {
  title: string;
  description?: string;
  status?: Status | string;
  dueDate?: string;
  priority?: Priority | string;
};

export async function createTask(task: TaskInput) {
  const response = await fetch("/api/tasks", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ task }),
    credentials: "include",
  });

  console.log(task);

  if (!response.ok) throw new Error("Failed to create task");

  window.location.reload();
}

export async function getTasks() {
  const response = await fetch("/api/tasks", {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
  });

  if (!response.ok) throw new Error("Failed to create task");

  const data = await response.json();

  return data;
}

export async function updateTask(taskId: string, task: TaskInput) {
  const response = await fetch("/api/tasks", {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ taskId, task }),
    credentials: "include",
  });

  if (!response.ok) throw new Error("Failed to create task");

  window.location.reload();
}

export async function deleteTask(taskId: string) {
  const response = await fetch("/api/tasks", {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ taskId }),
    credentials: "include",
  });

  if (!response.ok) throw new Error("Failed to create task");

  window.location.reload();
}
